packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -308,29 +321,41 @@ contract Bridge is EssentialContract, IBridge {
                     // time of the call.
                     //
                     // See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-150.md
-                    if (_message.gasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
 
-                    gasLimit = _message.gasLimit;
+                    if (invocationGasLimit > (gasleft() * 63) >> 6) revert B_NOT_ENOUGH_GASLEFT();
                 }
 
-                if (_invokeMessageCall(_message, msgHash, gasLimit)) {
+                if (_invokeMessageCall(_message, msgHash, invocationGasLimit)) {
                     _updateMessageStatus(msgHash, Status.DONE);
                 } else {
                     _updateMessageStatus(msgHash, Status.RETRIABLE);
                 }
             }
 
-            // Determine the refund recipient
-            address refundTo =
-                _message.refundTo == address(0) ? _message.destOwner : _message.refundTo;
-
             // Refund the processing fee
-            if (msg.sender == refundTo) {
-                refundTo.sendEtherAndVerify(_message.fee + refundAmount, _SEND_ETHER_GAS_LIMIT);
+            if (msg.sender == _message.destOwner) {
+                _message.destOwner.sendEtherAndVerify(
+                    _message.fee + refundAmount, _SEND_ETHER_GAS_LIMIT
+                );
             } else {
+                // gas now represents the actual gas used
+                uint256 fee;
+                {
+                    gasAmount -= gasleft();
+                    gasAmount += isNewlyProven
+                        ? ONE_STEP_PROCESSING_GAS_OVERHEAD
+                        : TWO_STEP_PROCESSING_GAS_OVERHEAD;
+
+                    uint256 maxFee = _message.fee * gasAmount / _message.gasLimit;
+                    uint256 actualFee = block.basefee * gasAmount / _message.gasLimit;
```

Comment:
```suggestion
                    uint256 actualFee = block.basefee * gasAmount;
```

I think this is the correct formula because basefee is per gas

Reply:
You are right, good catch.